java/debugger.jpda/src/org/netbeans/modules/debugger/jpda/models/ShortenedStrings.java

```diff
@@ -94,6 +95,42 @@ public static StringInfo getShortenedInfo(String s) {
         }
     }
 
+    private static boolean isSmallEndian(VirtualMachine virtualMachine) {
+        //TODO: Possibly cache this value
+        List<ReferenceType> possibleClasses = virtualMachine.classesByName(
+                "java.lang.StringUTF16");
```

Comment:
I think this should be cacheable with a WeakHashMap using the Virtualmachine as Key. This might be slow if the target VM is not local. I experimented a bit with this code and on JDK 17 initially `possibleClasses` was empty for me. Not sure how the VM handled this interernally, but I could make it work using this code:

```java
        if (possibleClasses.isEmpty()) {
            ClassType ct = (ClassType) virtualMachine.classesByName("java.lang.Class").iterator().next();
            Method m = ct.concreteMethodByName("forName", "(Ljava/lang/String;)Ljava/lang/Class;");
            StringReference referenceString = virtualMachine.mirrorOf("java.lang.StringUTF16");
            try {
                ThreadReference threadReference = virtualMachine.allThreads().get(0);
                ct.invokeMethod(threadReference, m, Collections.singletonList(referenceString), 0);
                possibleClasses = virtualMachine.classesByName("java.lang.StringUTF16");
            } catch (InvalidTypeException | ClassNotLoadedException | IncompatibleThreadStateException | InvocationException ex) {
                Exceptions.printStackTrace(ex);
            }
        }
```

Reply:
Aye. Will get to that. Do we need to handle multiple threads or are we good without using concurrent structures here?

Reply:
In my test I had to make `isSmallEndian` synchronized - else I got into a `IncompatibleThreadStateException`. Not really sure whether this hammer just fixed the problem accidentally.

Reply:
Ok. Should I slap a `synchronized` onto it, not worry about it, or find another solution?

Reply:
Works for me and looks sane. The caching you introduced works (at least in my trivial testcase).

One final idea would be to rework the synchronization. The maps are protected by individual `sychronized() {}` blocks. The pattern could be applied here two, which would also not leak the sychronization to the outside.

My suggestion is to move the `isLittleEndianCache.clear();` into a separate block `synchronized (isLittleEndianCache) {isLittleEndianCache.clear();}`, drop the `synchronized` from the signature of `isLittleEndian` and instead wrap the whole method contents into a `synchronized (isLittleEndianCache) {....}` block.